### PR TITLE
Convert some uses of ArrayList to List<T>

### PR DESCRIPTION
**Pull request by jkotas**
ArrayList.ToArray(Type) has to lookup the array type in the type loader structures. It makes it slow and potentially AOT unfriendly since the code required to support the array type may not exist.
Convert most uses of ArrayList.ToArray in libraries to use List<T> if possible, or ArrayList.CopyTo if the conversion to List<T> was not straightforward.

Simplified a few other places that created arrays while I was on it.

### CODE REVIEW

File: src/libraries/System.Configuration.ConfigurationManager/src/System/Configuration/ApplicationSettingsBase.cs [hunk 1]
```diff
@@ -771,7 +771,7 @@ internal static bool IsClickOnceDeployed(AppDomain appDomain)
         /// </summary>
         private PropertyInfo[] SettingsFilter(PropertyInfo[] allProps)
         {
-            ArrayList settingProps = new ArrayList();
+            List<PropertyInfo> settingProps = new List<PropertyInfo>();
```
Comment (davidfowl) [inline src/libraries/System.Configuration.ConfigurationManager/src/System/Configuration/ApplicationSettingsBase.cs:774]:
```suggestion
            var settingProps = new List<PropertyInfo>();
```

😃

File: src/libraries/System.Diagnostics.EventLog/src/System/Diagnostics/EventLogInternal.cs [hunk 2]
```diff
@@ -1413,7 +1413,7 @@ private sealed class LogListeningInfo
             public EventLogInternal handleOwner;
             public RegisteredWaitHandle registeredWaitHandle;
             public WaitHandle waitHandle;
-            public ArrayList listeningComponents = new ArrayList();
+            public List<EventLogInternal> listeningComponents = new List<EventLogInternal>();
```
Comment (davidfowl) [inline src/libraries/System.Diagnostics.EventLog/src/System/Diagnostics/EventLogInternal.cs:1416]:
```suggestion
            public List<EventLogInternal> listeningComponents = new();
```

File: src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/DirectoryServerCollection.cs [hunk 3]
```diff
@@ -14,7 +15,7 @@ public class DirectoryServerCollection : CollectionBase
         internal readonly DirectoryContext context;
         internal bool initialized;
         internal readonly Hashtable? changeList;
-        private readonly ArrayList _copyList = new ArrayList();
+        private readonly List<object> _copyList = new List<object>();
```
Comment (davidfowl) [inline src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/DirectoryServerCollection.cs:18]:
```suggestion
        private readonly List<object> _copyList = new();
```

Do we like target type new in dotnet/runtime? I haven't seen it being used much as yet.

Comment (jkotas) [reply src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/DirectoryServerCollection.cs]: I see a few hundred instances of it in newer code, so I guess we like it.

File: src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/DirectoryServerCollection.cs [hunk 4]
```diff
@@ -14,7 +15,7 @@ public class DirectoryServerCollection : CollectionBase
         internal readonly DirectoryContext context;
         internal bool initialized;
         internal readonly Hashtable? changeList;
```
Comment (davidfowl) [inline src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/DirectoryServerCollection.cs:17]: Dictionary?

File: src/libraries/System.Private.Xml/tests/XmlSchema/XmlSchemaValidatorApi/ExceptionVerifier.cs [hunk 5]
```diff
@@ -3,7 +3,7 @@
 
 using System.Linq;
 using Xunit.Abstractions;
-using System.Collections;
+using System.Collections.Generic;
```
Comment (stephentoub) [inline src/libraries/System.Private.Xml/tests/XmlSchema/XmlSchemaValidatorApi/ExceptionVerifier.cs:6]: The previous using is still necessary for Hashtable.

File: src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/DirectoryServerCollection.cs (hunk 4 above)
Comment (stephentoub) [reply src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/DirectoryServerCollection.cs:17]: The code is explicitly synchronizing access to this internal member.  It'd require review of all uses to ensure that was safe.  In general, replacing Hashtable with Dictionary needs review as they have subtly different thread-safety guarantees, and code frequently relies on Hashtable's, which enables lock-free reads with a single concurrent writer.

File: src/libraries/System.Private.Xml/tests/XmlSchema/XmlSchemaValidatorApi/ExceptionVerifier.cs (hunk 5 above)
Comment (stephentoub) [reply src/libraries/System.Private.Xml/tests/XmlSchema/XmlSchemaValidatorApi/ExceptionVerifier.cs:6]: (Same below)

**Review by stephentoub — APPROVED** [review]
LGTM other than the build failures.

File: src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/DirectoryServerCollection.cs (hunk 3 above)
Comment (stephentoub) [reply src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/DirectoryServerCollection.cs]: We're fine with it in cases like this where the type is still named as part of the operation. We're less happy with it when the code is just like:
```C#
somelocalorfield = new();
```
as that makes it much harder to understand at a glance what's happening.

File: src/libraries/System.Private.Xml/tests/XmlSchema/XmlSchemaValidatorApi/ExceptionVerifier.cs [hunk 6]
```diff
@@ -3,7 +3,7 @@
 
 using System.Linq;
 using Xunit.Abstractions;
```
Comment (jkotas) [inline src/libraries/System.Private.Xml/tests/XmlSchema/XmlSchemaValidatorApi/ExceptionVerifier.cs:5]:
```suggestion
using Xunit.Abstractions;
using System.Collections;
```

File: src/libraries/System.Private.Xml/tests/Xslt/XslTransformApi/ExceptionVerifier.cs [hunk 7]
```diff
@@ -3,7 +3,7 @@
 
 using System.Linq;
 using Xunit.Abstractions;
-using System.Collections;
+using System.Collections.Generic;
```
Comment (jkotas) [inline src/libraries/System.Private.Xml/tests/Xslt/XslTransformApi/ExceptionVerifier.cs:6]:
```suggestion
using System.Collections;
using System.Collections.Generic;
```

**Pull request merged**